sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ImplementIDisposableCorrectly.cs

```diff
@@ -48,203 +48,239 @@ public class ImplementIDisposableCorrectly : SonarDiagnosticAnalyzer
+            private readonly SemanticModel semanticModel;
+            private readonly List<SecondaryLocation> issues = new List<SecondaryLocation>();
+            private readonly INamedTypeSymbol classSymbol;
 
-            if (classSymbol == null || classSymbol.IsSealed)
+            public DisposableChecker(ClassDeclarationSyntax classDeclaration, SemanticModel semanticModel)
             {
-                return;
+                this.classDeclaration = classDeclaration;
+                this.semanticModel = semanticModel;
+                this.classSymbol = semanticModel.GetDeclaredSymbol(classDeclaration);
             }
 
-            var idisposableInterface = classSymbol.Interfaces.FirstOrDefault(IsOrImplementsIDisposable);
-            if (idisposableInterface == null)
+            public Diagnostic GetDiagnostic()
             {
-                return;
+                return Diagnostic.Create(rule, classDeclaration.Identifier.GetLocation(),
+                    issues.ToAdditionalLocations(),
+                    issues.ToProperties());
             }
 
-            var implementationErrors = new List<SecondaryLocation>();
-
-            if (classSymbol.BaseType.Implements(KnownType.System_IDisposable))
+            public bool IsCompliant()
             {
-                var idisposableInterfaceSyntax = classDeclaration.BaseList.Types
-                    .FirstOrDefault(t => IsOrImplementsIDisposable(t, c.SemanticModel));
-                if (idisposableInterfaceSyntax != null)
+                if (classSymbol == null || classSymbol.IsSealed)
                 {
-                    implementationErrors.Add(new SecondaryLocation(idisposableInterfaceSyntax.GetLocation(),
-                        $"Remove IDisposable from the list of interfaces implemented by {classSymbol.Name} and override the base class Dispose implementation instead."));
+                    return true;
                 }
-            }
-            else
-            {
-                if (!FindMethods(classSymbol, IsProtectedVirtualDisposeBool).Any())
+
+                if (classSymbol.BaseType.Implements(KnownType.System_IDisposable))
                 {
-                    implementationErrors.Add(new SecondaryLocation(classDeclaration.Identifier.GetLocation(),
-                        $"Provide protected overridable implementation of Dispose(bool) on {classSymbol.Name} or mark the type as sealed."));
+                    var idisposableInterfaceSyntax = classDeclaration.BaseList?.Types
+                        .FirstOrDefault(IsOrImplementsIDisposable);
+
+                    if (idisposableInterfaceSyntax != null)
+                    {
+                        AddLocation(idisposableInterfaceSyntax.GetLocation(),
+                            $"Remove IDisposable from the list of interfaces implemented by {classSymbol.Name}"
+                                + " and override the base class Dispose implementation instead.");

```

Comment:
This message should depend on whether the base class contains virtual Dispose(bool) or not